streams/shared/src/main/scala/zio/stream/ZSink.scala

```diff
@@ -1063,4 +1094,50 @@ object ZSink extends ZSinkPlatformSpecific {
       def stepPure(state: State, a: Any): Step[State, Nothing] = Step.done(state, Chunk.empty)
       def extractPure(state: State): Either[Nothing, B]        = Right(b)
     }
+
+  /**
+   * Creates a sink which throttles input elements of type A according to the given bandwidth parameters
+   * using the token bucket algorithm. The weight of each element is determined by the `costFn` effectful
+   * function. Stream elements are mapped to `Option[A]`, and `None` denotes that a given element has
+   * been throttled.
+   *
+   * The sink is returned as a managed resource in order to control the lifecycle of the fiber which
+   * concurrently refills the bucket with tokens.
+   */
+  final def throttle[R, E, A](
+    bucketCapacity: Long,
+    initialTokens: Long,
+    refill: Long,
+    refillInterval: Duration
+  )(
+    costFn: A => ZIO[R, E, Long]
+  ): ZManaged[R with Clock, E, ZSink[R, E, Nothing, A, Option[A]]] = {
+    import ZSink.internal._
+
+    def bucketSink(bucket: Ref[Long]) = new ZSink[R, E, Nothing, A, Option[A]] {
+      type State = (Ref[Long], Option[A])
+      val initial = UIO.succeed(Step.more((bucket, None)))
+      def step(state: State, a: A) =
+        for {
+          weight <- costFn(a)
+          result <- bucket.modify { tokens =>
+                     if (weight <= tokens) (Step.done((state._1, Some(a)), Chunk.empty), tokens - weight)
+                     else (Step.done((state._1, None), Chunk.empty), tokens)
+                   }
+        } yield result
+      def extract(state: State) = UIO.succeed(state._2)
+    }
+
+    def refillFn(tokens: Long): Long = Math.min(bucketCapacity, tokens + refill)
+
+    def refillBucket(bucket: Ref[Long]) =
+      bucket.update(refillFn).unit.delay(refillInterval).repeat(ZSchedule.spaced(refillInterval))
+
+    for {
+      _      <- ZManaged.fromEffect(assertBucketConfiguration(bucketCapacity, initialTokens, refill))
```

Comment:
The `ZManaged.fromEffect(x)` calls can be replaced by `x.toManaged_`

Reply:
Done.